File: leetcode/692.py
class Solution(object):
    def sortByOrder(a,b):
        return a[0] if a[1] == b[1] else a[1]
        
    def topKFrequent(self, words, k):
        """
        :type words: List[str]
        :type k: int
        :rtype: List[str]
        """
        dic = {}
        res = []
        for word in words:
            if word in dic.keys() : 
                dic[word] +=1
            else :
                dic[word] =0
        sortedDic = sorted(dic.items(), key = lambda x : (-x[1],x[0]))
        print sortedDic
        for i in range(k):
            res.append(sortedDic[i][0])
        return res


















# Solution 2 - on my own
# Using Dictionary, sort them with value and alphabetical order if values are the same
# beat 84.13%
# Learned to use lambda with sorted function 
# Time Complexity O(nLogn) : for using sort
class Solution(object):
    def topKFrequent(self, words, k):
        """
        :type words: List[str]
        :type k: int
        :rtype: List[str]
        """
	dic = {}
        for word in words:
                if word not in dic:
                        dic[word] = 0
                else :
                        dic[word]+=1
        res = sorted(dic.items(), key=lambda x:(-x[1],x[0]))
        res = res[0:k]
        word = []
        for tup in res:
                word.append(tup[0])
        return word

File: leetcode/692.cpp
// 692. Top K Frequent Words
// Solution 1
// Brute-Force approach
// First, using map, already sort alphabetically.
// Second, Find max Value and insert them 
// Time Complexity : O(N^N)
class Solution {
public:
 
    vector<string> topKFrequent(vector<string>& words, int k) {
        
        vector<string> res;
        if(words.empty())
            return res;
        
        map<string,int> m;
        for (int i =0; i< words.size(); i++){
            if(m.count(words[i]) < 0)
                m[words[i]]=0;
            else
                m[words[i]]++;
        }
        // need to sort by value
        for(int i=0; i<k; i++){
            int max = 0; auto temp = m.begin();
            for(auto itr = m.begin();itr!=m.end();itr++ ){
                if(max < itr->second){
                    max = itr->second;
                    temp = itr;
                }
            }
            res.push_back(temp->first);
            m[temp->first] = 0;
        }
        
        return res;
    }
};
